packages/flutter/lib/src/widgets/app_lifecycle_listener.dart

```diff
@@ -0,0 +1,141 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+/// A callback type that is used by [AppLifecycleListener.onExitRequested] to
+/// ask the application if it wants to cancel application termination or not.
+typedef AppExitRequestCallback = Future<AppExitResponse> Function();
+
+/// A listener that can be used to listen to changes in the application
+/// lifecycle.
+///
+/// To listen for requests for the application to exit, and to decide whether or
+/// not the application should exit when requested, create an
+/// [AppLifecycleListener] and set the [onExitRequested] callback. On macOS, in
+/// order to receive application exit requests originating from the OS, the
+/// `NSPrincipalClass` in your `Info.plist` file must be set to
+/// `FlutterApplication`, or a subclass of `FlutterApplication`.
+///
+/// To listen for changes in the application lifecycle state, define an
+/// [onStateChange] callback. The most recent state seen by this
+/// [AppLifecycleListener] can be retrieved with the [lifecycleState] accessor.
+/// See the [AppLifecycleState] enum for details on the various states.
+///
+/// {@tool dartpad}
+/// This examples shows how an application can optionally decide
+/// to abort a request for exiting instead of obeying the request.
+///
+/// ** See code in examples/api/lib/widgets/app_lifecycle_listener/app_lifecycle_listener.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [ServicesBinding.exitApplication] for a function to call that will request
+///   that the application exits.
+/// * [WidgetsBindingObserver.didRequestAppExit] for the handler which this
+///   class uses to receive exit requests.
+/// * [WidgetsBindingObserver.didChangeAppLifecycleState] for the handler which
+///   this class uses to receive lifecycle state changes.
+class AppLifecycleListener with WidgetsBindingObserver, Diagnosticable {
```

Comment:
This seems to be a thin wrapper around WidgetsBindingObserver. Do we think that this will be such a common use case that this wrapper provides much value? What makes the lifecycle stuff special compared to the other functionality in WidgetsBindingObserver that is not getting this special treatment?